### ISSUE TITLE
Support pub packages

### ISSUE THREAD

**Issue by lukechurch** [dart-lang/dartpad_server#43]
@devoncarew has started this

https://github.com/dart-lang/dartpad_server/pull/40

We need to consider:

-> Ensure that it can't be used as a DOS vector by triggering on the expensive package resolutions
-> Ensure that it plays nicely with intermittent failures of pub storage

**Comment by devoncarew** [comment]
We have some library support for this now, but nothing exposed via the REST API. We are doing validation/normalization of the package names and have a timeout set to mitigate pathological dep. resolution times.

**Comment by MikeMitterer** [comment]
He guys, how is the status on this? DartPad could be THE tool to show the world what can be done with Dart but without packages its just a little toy...

**Comment by lukechurch** [comment]
Hi @MikeMitterer, DartPad is a little bit in a holding pattern right now whilst we're doing work elsewhere in Dart.

The easiest thing to do here would probably to add a whitelist of a bunch of packages. Do you have any specifically in mind you'd want to use?

**Comment by kasperpeulen** [comment]
@lukechurch I now that there is interest from angular.io for showing angular examples using something like DartPad. But I guess that would even more difficult, because it would also need multi file support. @kwalrath

**Comment by zoechi** [comment]
Multi-file support would be nice but shouldn't be necessary for Angular. You could put everything in `main.dart` (but not very convenient).
Angular has specific `pubspec.yaml` requirements (transformer and transformer options).
Angular and Polymer/polymer_elements support would be great though :)

**Comment by marcojakob** [comment]
@lukechurch I'm sad to hear that DartPad is on hold. I was hoping so much that DartPad could become a replacement for DartEditor in most use cases (for me it's mostly educational purposes). But without general package support and in a lot of cases even multi-file support, DartPad is really not usable for me.

This is sad, because I feel that we're losing (or not winning) a large group of potential Dart users that just can't work with something like Webstorm or even Atom. Those editors including the Dart SDK are too difficult to manage and maintain for a school context.

**Comment by MikeMitterer** [comment]
@kasperpeulen + @lukechurch I'd like to show some samples from mdl-package. e.g. http://styleguide.material-design-lite.pub/#/button multi-file support is not necessary for mdl! If you click on "Basic usage" you see what I mean.

I think DartPad should have high priority - maybe it was not so frequently used in the past but this was (in my opinion) just because of the missing package support!

**Comment by zoechi** [comment]
I don't expect Dartpad to be used for real-world development (but I'm not against it or course) but for support and sharing examples it's invaluable. 
I'm currently learning Angular2 and there are tons of Plunker links shared everywhere for TS but there is no way to do the same with Angular2 for Dart :-(

**Comment by lukechurch** [comment]
@marcojakob Priorities change quickly, so whilst it's not something we're working on this week - that doesn't mean it won't get more attention in the near future. Your needs sound fairly achievable. I remember our discussion from before, I would really like to support the education use case more strongly. I will poke at whether there are more creative ways of getting the work done. If you have any more data you can share about your instinct that we're losing potential Dart users due to the deployment issues, I'd _really_ appreciate hearing about it, I know this is on  @filiph 's mind as well.

The multiple-files aspect is primarily UI work, the dart-services component already largely understands how to handle multiple files. @kasperpeulen don't suppose you have any free UI cycles do you? :)

Some of the work I'm doing elsewhere is teaching me a lot about reliably handling arbitrary pubspecs, that may well propagate into an easy fix for the pub issues.

@MikeMitterer are you up for doing an experiment if I pass you a test version which has your package added to the supported packages? Are you looking at trying to host it via embedding in your site or via samples in a DartPad-like site?

**Comment by marcojakob** [comment]
@lukechurch More data on **"my instinct for potential Dart users"**:
I've been proposing Dart as *the best programming language for education*™ for almost three years. I wrote an article (see [this blog post in German](http://code.makery.ch/blog/web-und-mobile-mit-dart/)) and gave workshops at the largest conference for computer science in education in the German speaking part of Europe in 2013 (called INFOS). The people were excited about what you can do with Dart.

The *best* language for education has the following features:

1. Makes programming novices happy.
2. Makes professional programmers happy and productive.
3. Runs on mobile platforms (as web application is acceptable).
4. Runs on a desktop: Windows, MacOS, Linux
5. Has a broad spectrum of possible applications: interactive games, animations, business applications, client-server applications, etc.

Still, even after 3 years, I would say that Dart is the only programming language to sufficiently fulfill all those requirements. I could go on about why it's way better suited for education than JavaScript (and all its flavors), better than Java and better than Python (which are probably still the most popular languages used in education). I wrote some arguments in the article mentioned above and on [this page](http://code.makery.ch/library/hello-dart/de/background/) (sorry, currently in German only).

I started writing [multiple tutorial series](http://code.makery.ch/library/hello-dart/) that could be used to learn Dart. Then I stopped all my effort because the DartEditor was shut down. And no, Webstorm is not an option because you have to get a license, then install the SDK, then install Webstorm to get a very complicated IDE with a difficult to use debugger. Similar for the other editor options. Sorry, but DartEditor was perfect in that regard.

Now, think about it: You have a wonderful language and an ecosystem around it. Innovative people in education are waiting for a language with exactly those features. If you could provide the tools and resources and maybe some training for it, you could get Dart into schools and the kids and teachers would love it. Here lies the potential to win the future generation and help them write beautiful apps with an awesome language!

**Comment by stevenroose** [comment]
@marcojakob Regarding your feelings about the DartEditor: Consider the Chrome Dev Editor as an alternative:
https://chrome.google.com/webstore/detail/chrome-dev-editor-develop/pnoffddplpippgcfjdhbmhkofpnaalpg?hl=en

**Comment by marcojakob** [comment]
@stevenroose Development for Chrome Dev Editor has also [been stopped](https://plus.google.com/+SriramSaroop/posts/6EwgknwKpeS).

**Comment by zoechi** [comment]
Atom is IMHO the closest to DartEditor currently.

**Comment by MikeMitterer** [comment]
@marcojakob Is the license really a problem? Students should get it for free. In my opinion students should also be able to use a major IDE - this is part of the whole process. It's not necessary to use the debugger at the beginning of a programmer career  and after a few weeks/month they are (hopefully) experienced enough to handle the debugger.

**Comment by kasperpeulen** [comment]
@marcojakob I can't agree more with this. The major reason I've been so excited about DartPad, is because I feel Dart is the ultimate language for education. 

> The multiple-files aspect is primarily UI work, the dart-services component already largely understands how to handle multiple files. @kasperpeulen don't suppose you have any free UI cycles do you? :)

How do you picture this? Like some kind of fork of DartPad? Or in DartPad itself.

**Comment by lukechurch** [comment]
@kasperpeulen: Right now what I'd propose is a little '+' button next to the files list. When you click this it offers to add a new Dart file, and renames the existing 'Dart' tab to 'main.dart'. So in Dartpad itself. Does this make sense?

At this stage we'd need to do the experiments in a fork of DartPad just so that we don't interrupt the current users until we were really ready. We have test projects we can run that from etc... So there are no issues there.

**Comment by marcojakob** [comment]
@MikeMitterer Yes, the license is a problem. It's true that students can get it for free, but the problem is that it's one additional administrative step that teachers will have to take. And to do this they would need to be quite convinced about Dart already which they aren't yet, because they don't know what they're missing. Then you need admin rights to install Webstorm (which isn't easy to get in some schools), you must manually update the SDK,... (DartEditor was unzip, no-install, and autoupdating). Then there are the teachers that are very picky about commercial products in general...

**Comment by stevenroose** [comment]
@marcojakob FWIW, IntelliJ products do not require admin rights on Linux. Especially now that they ship with a JDK inside (previously admin rights were required for java installation).

Just unzip and run. I suppose for Windows a similar this must be possible, although most probably more difficult (since it would involve running the `.jar` manually).

Also, IDEA Community Edition also has a Dart plugin IIRC..

**Comment by MikeMitterer** [comment]
@marcojakob Got you + agree - the barrier must be as low as possible but as @stevenroose said - IntelliJ is also quite easy to install. In my opinion the problem is more the installer under windows than the IDE - Chocolatey is nice if you like the cmdline but if not, and most students don't event know how to get to the cmdline, then there is only this "using a third-party installer" option on the Dart website. This installer should be the number one and it should be maintained by the Dart Team! Node.js does it right: https://nodejs.org/en/download/

**Comment by hermanbergwerf** [comment]
Yeah, just tried to build a little Dart example in WebGL from https://github.com/BoldInventions/dart-webgl-tutorials/tree/master/web/lesson_04, but it uses `vector_math` :-(

**Comment by davidmorgan** [comment]
Any update please?

I'd love to be able to point to demos of my packages on dartpad. Even simple support would be good -- but of course, full package support would be great :)

**Comment by szepeshazi** [comment]
Any update please?

**Comment by stephenhuey** [comment]
Why not load a whitelist?  For example:  package:meta/meta.dart
